clang-tools-extra/clang-tidy/modernize/UseNullptrCheck.cpp

```diff
@@ -38,7 +38,9 @@ AST_MATCHER(Type, sugaredNullptrType) {
 StatementMatcher makeCastSequenceMatcher(llvm::ArrayRef<StringRef> NameList) {
   auto ImplicitCastToNull = implicitCastExpr(
       anyOf(hasCastKind(CK_NullToPointer), hasCastKind(CK_NullToMemberPointer)),
-      unless(hasImplicitDestinationType(qualType(substTemplateTypeParmType()))),
+      anyOf(hasSourceExpression(gnuNullExpr()),
```

Comment:
release note mentions this patch will improve `NULL` and `__null`. But here I can only find `__null`. Is there any mismatch?

Reply:
This is explained somewhat in the unit test: When `NULL` is defined to `__null`, then this change will catch both cases. But we don't have a direct check for "the origin was the macro `NULL`", I'm afraid. @zygoloid suggested this change as a reasonable, small change that will be effective in many real-world cases.

I'd be happy to have a more general solution, but couldn't find one, and the present one will already do a lot of good. What do you think?

Reply:
I agree the current solution, could you add some comment TODO or FIXME to simplify troubleshooting when this issue is reported。

Reply:
Sure -- do you mean here in the implementation, or elsewhere?

Reply:
I prefer in the `makeCastSequenceMatcher`

Reply:
How about:

```
/// TODO/NOTE: The second "anyOf" below discards matches on a substituted type,
/// since we don't know if that would _always_ be a pointer type for all other
/// specializations, unless the expression was "__null", in which case we assume
/// that all specializations are expected to be for pointer types. Ideally this
/// would check for the NULL macro instead, but that'd be harder to express. In
/// practice, NULL is often defined as __null, and this is a useful condition.
```

?

Reply:
Done, PTAL!